src/NuGet.Core/NuGet.Packaging/NupkgMetadata/NupkgMetadataFileFormat.cs

```diff
@@ -31,12 +41,35 @@ public static NupkgMetadataFile Read(string filePath, ILogger log)
 
         public static NupkgMetadataFile Read(Stream stream, ILogger log, string path)
         {
-            using (var textReader = new StreamReader(stream))
+            if (stream is null) { throw new ArgumentNullException(nameof(stream)); }
+
+            try
             {
-                return Read(textReader, log, path);
+                try
+                {
+                    NupkgMetadataFile nupkgMetadata = System.Text.Json.JsonSerializer.Deserialize<NupkgMetadataFile>(stream, JsonSerializerOptions);
+                    if (nupkgMetadata == null)
+                    {
+                        throw new InvalidDataException();
+                    }
+                    return nupkgMetadata;
+                }
+                catch (System.Text.Json.JsonException jsonException)
+                {
+                    throw new InvalidDataException(jsonException.Message, jsonException);
+                }
+            }
+            catch (Exception ex)
+            {
+                log.LogWarning(string.Format(CultureInfo.CurrentCulture,
+                    Strings.Error_LoadingHashFile,
+                    path, ex.Message));
+
+                throw;
             }
         }
 
+        [Obsolete("Use a different overload for better performance. This overload will get deleted in the future.")]
         public static NupkgMetadataFile Read(TextReader reader, ILogger log, string path)
```

Comment:
We could be more aggressive and just delete this public API, as well as the `Write(TextWriter` overload. That would allow us to delete all the Newtonsoft.Json code from this class.